### PR TITLE
feat: communities upgrades

### PR DESCRIPTION
**Pull request by abiudrn**
- add image upload logic
- view images using cachednetwork images
- update endpoints to use prohealth
- remove images storage from state
- add last since to the sync observer
- remove uint8list converter from utils
- extract and refactor message input
- refactor view models to only use what is required
- reduce sync interval to 3000ms
- add selected room to state
- add backoff implementation in sync action to address errors
- add set room action
- add set backoff action to handle breakages in sync
- add reject invite logic
- add promote to moderator logic
- update endpoints to add moderator endpoint
- refactor utils
- add online members in room data class
- add group avatars
- add preview image functionality
- add avatars in all user display screens
- refactor view models for simplicity
- fetch members after inviting someone
- refactoe sync action for simplicity
- refactor avatar widget
- add image preview when uploading
- add refresh indicator
- fix room list view model
- fix user list item widget
- fix UI in room page when an invite has been received
- fix failing tests

### CODE REVIEW

File: lib/presentation/communities/invited_groups/pages/image_preview_page.dart [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                        strokeWidth: 1,
+                      ),
+                    ),
+                  );
+                },
+              ),
+            ),
+            mediumVerticalSizedBox,
+            const Text(
+              'Image details',
```
Comment (abiudrn) [inline lib/presentation/communities/invited_groups/pages/image_preview_page.dart:60]: Extract these strings

File: lib/presentation/communities/invited_groups/pages/normal_message_widget.dart [hunk 2]
```diff
@@ -52,15 +50,22 @@ class NormalMessageWidget extends StatelessWidget {
...
-          url: imageUrl,
+          imageURL: imgURI,
         );
 
+      // case 'm.video':
+      //   return VideoPlayer(content['url']);
+
+      // case MatrixMessageTypes.file:
+      //   final String fileName = messageContent['filename']?.toString() ?? '';
+      //   return Text('This is a file: $fileName');
```
Comment (abiudrn) [inline lib/presentation/communities/invited_groups/pages/normal_message_widget.dart:68]: Remove dead code

File: lib/presentation/communities/widgets/avatar.dart [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        height: aviSize,
+        width: aviSize,
+        decoration: BoxDecoration(
+          shape: BoxShape.circle,
+          color: Colors.purple.withOpacity(0.2),
+        ),
+        padding: const EdgeInsets.all(15),
+        child: Center(
+          child: Text(
+            getInitials(displayName ?? 'No name').toUpperCase(),
```
Comment (abiudrn) [inline lib/presentation/communities/widgets/avatar.dart:61]: Extract this string

File: lib/application/redux/actions/communities/promote_to_moderator_action.dart [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    if (response.statusCode == 408) {
+      /// handle errors here
+
+    }
+
+    final Map<String, dynamic> decodedResponse =
+        json.decode(response.body) as Map<String, dynamic>;
+
+    if (decodedResponse['errcode'] != null) {
+      // Handle errors here
```
Comment (byronkimani) [inline lib/application/redux/actions/communities/promote_to_moderator_action.dart:58]: Add TODO here

File: lib/presentation/communities/invited_groups/widgets/send_message_input.dart [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                        if (msgCtrl.text.isNotEmpty) {
+                          StoreProvider.dispatch<AppState>(
+                            context,
+                            SendMessageAction(
+                              roomID: roomID,
+                              onSuccess: (String roomID) {
+                                if (roomID.isNotEmpty) {
+                                  msgCtrl.clear();
+                                } else {
+                                  // Show an error snackbar here
```
Comment (byronkimani) [inline lib/presentation/communities/invited_groups/widgets/send_message_input.dart:65]: Add TODO here

File: lib/presentation/communities/invited_groups/widgets/send_message_input.dart [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                if (msgCtrl.text.isNotEmpty) {
+                  StoreProvider.dispatch<AppState>(
+                    context,
+                    SendMessageAction(
+                      roomID: roomID,
+                      onSuccess: (String roomID) {
+                        if (roomID.isNotEmpty) {
+                          msgCtrl.clear();
+                        } else {
+                          // Show an error snackbar here
```
Comment (byronkimani) [inline lib/presentation/communities/invited_groups/widgets/send_message_input.dart:99]: Add TODO here

**Review by byronkimani — CHANGES_REQUESTED** [review]

**Review by byronkimani — APPROVED** [review]

**Review by EugeneDevv — APPROVED** [review]

**Pull request merged**